planner/core/rule_join_reorder.go

```diff
@@ -33,31 +33,64 @@ import (
 		// For example, the join type is cross join or straight join, or exists the join algorithm hint, etc.
 		// We need to return the hint information to warn
-		hintInfo = append(hintInfo, join.hintInfo)
+		joinOrderHintInfo = append(joinOrderHintInfo, join.hintInfo)
+	}
+	// `leftHasHint` and `rightHasHint` are used to record whether the left child and right child are set by the join method hint.
+	leftHasHint, rightHasHint := false, false
+	if isJoin && p.SCtx().GetSessionVars().EnableAdvancedJoinHint && join.preferJoinType > uint(0) {
+		// If the current join node has the join method hint, we should store the hint information and restore it when we have finished the join reorder process.
+		if join.leftPreferJoinType > uint(0) {
+			joinMethodHintInfo[join.children[0].ID()] = &joinMethodHint{join.leftPreferJoinType, join.hintInfo}
+			leftHasHint = true
+		}
+		if join.rightPreferJoinType > uint(0) {
+			joinMethodHintInfo[join.children[1].ID()] = &joinMethodHint{join.rightPreferJoinType, join.hintInfo}
+			rightHasHint = true
+		}
 	}
-	if !isJoin || join.preferJoinType > uint(0) || join.StraightJoin ||
+	// If the variable `tidb_opt_advanced_join_hint` is false and the join node has the join method hint, we will not split the current join node to join reorder process.
+	if !isJoin || (join.preferJoinType > uint(0) && !p.SCtx().GetSessionVars().EnableAdvancedJoinHint) || join.StraightJoin ||
 		(join.JoinType != InnerJoin && join.JoinType != LeftOuterJoin && join.JoinType != RightOuterJoin) ||
 		((join.JoinType == LeftOuterJoin || join.JoinType == RightOuterJoin) && join.EqualConditions == nil) {
-		if hintInfo != nil {
+		if joinOrderHintInfo != nil {
 			// The leading hint can not work for some reasons. So clear it in the join node.
 			join.hintInfo = nil
 		}
-		return []LogicalPlan{p}, nil, nil, nil, hintInfo, false
+		return &joinGroupResult{
+			group:             []LogicalPlan{p},
+			joinOrderHintInfo: joinOrderHintInfo,
+			basicJoinInfo:     &basicJoinInfo{},
+		}
 	}
 	// If the session var is set to off, we will still reject the outer joins.
 	if !p.SCtx().GetSessionVars().EnableOuterJoinReorder && (join.JoinType == LeftOuterJoin || join.JoinType == RightOuterJoin) {
-		return []LogicalPlan{p}, nil, nil, nil, hintInfo, false
+		return &joinGroupResult{
+			group:             []LogicalPlan{p},
+			joinOrderHintInfo: joinOrderHintInfo,
+			basicJoinInfo:     &basicJoinInfo{},
+		}
 	}
 	hasOuterJoin = hasOuterJoin || (join.JoinType != InnerJoin)
-	if join.JoinType != RightOuterJoin {
-		lhsGroup, lhsEqualConds, lhsOtherConds, lhsJoinTypes, lhsHintInfo, lhsHasOuterJoin := extractJoinGroup(join.children[0])
+	// If the left child has the hint, it means there are some join method hints want to specify the join method based on the left child.
+	// So we don't need to split the left child part. The right child part is the same, so omit the comments for the right child part.
+	if join.JoinType != RightOuterJoin && !leftHasHint {
```

Comment:
1. I don't understand why join type hint would block the join reorder process.
2. Please add a comment to reference here in the right child part. If someone sees the right child part code first, he may not notice the comments.